daemon/device-plugin/deviceplugin.go

```diff
@@ -0,0 +1,243 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cnitypes"
+	pb "github.com/openshift/dpu-operator/tree/main/dpu-api/gen"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/types/known/emptypb"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	VendorPluginSocketPath string = cnitypes.DaemonBaseDir + "vendor-plugin/vendor-plugin.sock"
+
+	// Device plugin settings.
+	pluginMountPath = "/var/lib/kubelet/device-plugins"
+	kubeletEndpoint = "kubelet.sock"
+	pluginEndpoint  = "sriovNet.sock"
+	resourceName    = "intel.com/ipu"
+	prefix          = "intel.com"
+)
+
+// sriovManager manages sriov networking devices
+type apfManager struct {
+	socketFile string
+	devices    map[string]pluginapi.Device // for Kubelet DP API
+	grpcServer *grpc.Server
+	pluginapi.DevicePluginServer
+	log    logr.Logger
+	client pb.DeviceServiceClient
+	conn   *grpc.ClientConn
+}
+
+type DevicePlugin interface {
+	Start() error
+}
+
+func (sm *apfManager) ListAndWatch(empty *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	changed := true
+	for {
+		for id, dev := range sm.devices {
+			state := sm.GetDeviceState(id)
+			if dev.Health != state {
+				changed = true
+				dev.Health = state
+				sm.devices[id] = dev
+			}
+		}
+		if changed {
+			resp := new(pluginapi.ListAndWatchResponse)
+			for _, dev := range sm.devices {
+				resp.Devices = append(resp.Devices, &pluginapi.Device{ID: dev.ID, Health: dev.Health})
+			}
+			fmt.Printf("ListAndWatch: send devices %v\n", resp)
+			if err := stream.Send(resp); err != nil {
+				fmt.Printf("Error. Cannot update device states: %v\n", err)
+				sm.grpcServer.Stop()
+				return err
+			}
+		}
+		changed = false
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// Allocate passes the dev name as an env variable to the requesting container
+func (sm *apfManager) Allocate(ctx context.Context, rqt *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	resp := new(pluginapi.AllocateResponse)
+	devName := ""
+	for _, container := range rqt.ContainerRequests {
+		containerResp := new(pluginapi.ContainerAllocateResponse)
+		for _, id := range container.DevicesIDs {
+			fmt.Printf("DeviceID in Allocate: %v \n", id)
+			dev, ok := sm.devices[id]
+			if !ok {
+				fmt.Printf("Error. Invalid allocation request with non-existing device %s", id)
+			}
+			if dev.Health != pluginapi.Healthy {
+				fmt.Printf("Error. Invalid allocation request with unhealthy device %s", id)
+			}
+
+			devName = devName + id + ","
+		}
+
+		fmt.Printf("device(s) allocated: %s\n", devName)
+		envmap := make(map[string]string)
+		envmap["APF-DEV"] = devName
```

Comment:
We should Get devices by through the "Resolve" API that would be implemented by the VSP.
So we should do  ports = Resolve(POD_UUID, VLAN ID (a.k.a chain ID) that would store this data in the VSP and return back the ports (APF interfaces) in the following struct : 
Name str;
PortType enum; (PortType can be NetDev or PCIEDev or VirtualIf - Standard Linux Interface Constructs)

This would be implemented in the NF CNIServer code. Let me know if you want me to do a prototype writeup for this.

Reply:
Let's park this for now. We will revisit it in 3 weeks

Reply:
Agreed